flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/harness/CountTriggerWithCleanupStateHarnessTest.scala

```diff
@@ -0,0 +1,305 @@
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow
+import org.apache.flink.streaming.runtime.operators.windowing.TriggerTestHarness
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord
+import org.apache.flink.table.api.StreamQueryConfig
+import org.apache.flink.table.runtime.triggers.CountTriggerWithCleanupState
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class CountTriggerWithCleanupStateHarnessTest {
+  protected var queryConfig =
+    new StreamQueryConfig().withIdleStateRetentionTime(Time.seconds(2), Time.seconds(3))
+
+  @Test
+  def testFiringAndFireingWithPurging(): Unit = {
+    val testHarness = new TriggerTestHarness[Any, TimeWindow](
+      CountTriggerWithCleanupState.of[TimeWindow](queryConfig, 10), new TimeWindow.Serializer)
+
+    // try to trigger onProcessingTime method via 1, but there is non timer is triggered
+    assertEquals(0, testHarness.advanceProcessingTime(1).size())
+
+    // register cleanup timer with 3001
+    assertEquals(
+      TriggerResult.CONTINUE,
+      testHarness.processElement(new StreamRecord[Any](1), new TimeWindow(0, 9)))
+
+    // try to trigger onProcessingTime method via 1000, but there is non timer is triggered
+    assertEquals(0, testHarness.advanceProcessingTime(1000).size())
+
+    // 1000 + 2000 <= 3001 reuse timer 3001
+    assertEquals(
+      TriggerResult.CONTINUE,
+      testHarness.processElement(new StreamRecord[Any](1), new TimeWindow(0, 9)))
+
+    // there are two state entries, one is timer(3001) another is counter(2)
+    assertEquals(2, testHarness.numStateEntries)
+
+    // try to trigger onProcessingTime method via 3001, and timer(3001) is triggered
+    assertEquals(
+      TriggerResult.FIRE_AND_PURGE,
+      testHarness.advanceProcessingTime(3001).iterator().next().f1)
+
+    assertEquals(0, testHarness.numStateEntries)
+
+    // 3001 + 2000 >= 3001 register cleanup timer with 6001, and remove timer 3001
+    assertEquals(
+      TriggerResult.CONTINUE,
+      testHarness.processElement(new StreamRecord[Any](1), new TimeWindow(0, 9)))
+
+    // try to trigger onProcessingTime method via 4002, but there is non timer is triggered
+    assertEquals(0, testHarness.advanceProcessingTime(4002).size())
+
+    // 4002 + 2000 >= 6001 register cleanup timer via 7002, and remove timer 6001
+    assertEquals(
+      TriggerResult.CONTINUE,
+      testHarness.processElement(new StreamRecord[Any](1), new TimeWindow(0, 9)))
+
+    // 4002 + 2000 <= 7002 reuse timer 7002
+    assertEquals(
+      TriggerResult.CONTINUE,
+      testHarness.processElement(new StreamRecord[Any](1), new TimeWindow(0, 9)))
+
+    // have one timer 7002
+    assertEquals(1, testHarness.numProcessingTimeTimers)
+    assertEquals(0, testHarness.numEventTimeTimers)
+    assertEquals(2, testHarness.numStateEntries)
+    assertEquals(2, testHarness.numStateEntries(new TimeWindow(0, 9)))
+    assertEquals(0, testHarness.numStateEntries(new TimeWindow(9, 18)))
```

Comment:
There is only a single global window. So, we do not need to test for different windows, IMO.